Source/JavaWrappers.cpp

```diff
@@ -274,6 +281,88 @@ namespace java::lang
     }
 }
 
+namespace java::websocket
+{
+    WebSocketClient::WebSocketClient(std::string url, std::function<void(std::string)> message, std::function<void()> open, std::function<void()> close, std::function<void()> error)
+        : Object{android::global::GetWebSocketClass()}
+        , m_openCallback{std::move(open)}
+        , m_messageCallback{std::move(message)}
+        , m_closeCallback{std::move(close)}
+        , m_errorCallback{std::move(error)}
+    {
+        static JNINativeMethod methods[] = {
+                {"closeCallback", "()V", (void*)OnClose },
+                {"openCallback", "()V", (void*)&WebSocketClient::OnOpen },
+                {"messageCallback", "(Ljava/lang/String;)V", (void*)&WebSocketClient::OnMessage },
+                {"errorCallback", "()V", (void*)&WebSocketClient::OnError },
+        };
+        m_env->RegisterNatives(m_class, methods, 4);
+
+        // initialize socket
+        JObject(m_env->NewObject(m_class, m_env->GetMethodID(m_class, "<init>", "(Ljava/lang/String;)V"), m_env->NewStringUTF(url.c_str())));
+
+        WebSocketClient *p_webSocketClient = this;
+        s_objectMap.insert({JObject(), p_webSocketClient});
+
+        jmethodID connectSocket{m_env->GetMethodID(m_class, "connectBlocking", "()Z")};
+        m_env->CallBooleanMethod(JObject(), connectSocket);
+    }
+
+    void WebSocketClient::OnOpen(JNIEnv* env, jobject obj) {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) {
+            itObject->second->m_openCallback();
+        }
+    }
+
+    void WebSocketClient::OnMessage(JNIEnv* env, jobject obj, jstring message) {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) {
+            java::lang::String mystr{message};
+            itObject->second->m_messageCallback(mystr);
+        }
+    }
+
+    void WebSocketClient::OnClose(JNIEnv* env, jobject obj) {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) {
+            itObject->second->m_closeCallback();
+
+            // delete the websocket object from map
+            WebSocketClient::s_objectMap.erase(itObject);
+        }
+    }
+
+    void WebSocketClient::OnError(JNIEnv* env, jobject obj) {
+        auto itObject = WebSocketClient::FindClientInstance(env, obj);
+        if (itObject != WebSocketClient::s_objectMap.end()) {
+            itObject->second->m_errorCallback();
+        }
+    }
+
+    std::unordered_map<jobject, java::websocket::WebSocketClient*>::iterator  WebSocketClient::FindClientInstance(JNIEnv* env, jobject obj) {
```

Comment:
>std::unordered_map<jobject, java::websocket::WebSocketClient*>::iterator [](http://example.com/codeflow?start=4&length=72)

this seems kinda awkward, why not just return WebSocketClient*?  From all the callsites you're just accessing the client pointer.

If the concern is returning something for the not-found case, you can either return an nullptr for the caller to check, or return a bool to signify success/failure and have the actual client be an outparam.

Reply:
Wouldn't the iterator only return the MapIterator instead which implies to be a pointer to a position inside the Map. 
If I'm not mistaken, you can access the Iterator helpers to get the ```value```

Reply:
I see, in that case you're not actually using an unordered_map, right?  You're effectively using it as a std::vector as you're doing an O(N) lookup.  And if that's the case, changing the type to std::vector would actually clear a lot of this up in terms of coding intent.

Reply:
Spoke offline, switched to std::vector.